L52-objc-bazel-support.md

```diff
@@ -0,0 +1,124 @@
+
+We use the native `proto_library` rule as a manager for `.proto` files (i.e. their package paths and dependencies). They wrap the proto files and are passed into `objc_proto_lib` and `objc_grpc_lib` as `srcs`. As a result of using `proto_library`, proto files must not reside in a upper-level directory to the `BUILD` file.
+
+Due to naming rules within the protobuf compiler, we first need to convert file names into camel case in order to correctly declare all the output files. File extensions are either `.pbobjc.h/.m` or `.pbrpc.h/.m`, give the pair of files that will be compiled (Messages or Services, respectively) as a boolean parameter.
+
+The ObjC plugin is one of the arguments. The precise arguments is formatted by the libary function `get_plugin_args` provided in `//bazel:protobuf.bzl`.
+
+If the proto files are depend on any of the "well known" proto files like `empty.proto` defined in `@com_google_protobuf//google/protobuf/`, this external rule `@com_google_protobuf//:well_known_protos` is needed in order to add those files into arguments.
+
+Outputs will be generated in directories managed by Bazel. They will be further splitted into two groups - one for `hdrs` and the other for `srcs`/`non_arc_srcs` - and then passed into a native `objc_library` rule for the final compiled library. Note that one `objc_proto/grpc_lib` might need to depend on others as well.
+
+### Example
+
+Suppose a very simple HelloWorldClient project with a single proto file `//HelloWorld/hello.proto` with both messages and services. Assume that it depends on one of the well known protos.
+
+Configure `WORKSPACE` as instructed in this [tutorial](https://docs.bazel.build/versions/master/tutorial/ios-app.html) and also load `grpc_deps`:
+```
+git_repository(
+    name = "com_github_grpc_grpc",
+    remote = "https://github.com/grpc/grpc.git",
+    branch = "master"
+)
+load("@com_github_grpc_grpc//bazel:grpc_deps.bzl", "grpc_deps")
+grpc_deps()
+```
+
+And here is the `BUILD` file:
+```
+load("@build_bazel_rules_apple//apple:ios.bzl", "ios_application")
+load(
+    "@com_github_grpc_grpc//bazel:grpc_objc_library.bzl", 
+    "objc_proto_lib", 
+    "objc_grpc_lib",
+)
+proto_library(
+    name = "hello_proto",
+    srcs = ["HelloWorld/hello.proto"]
+)
+objc_proto_lib(
+    name = "hello_proto_objc",
+    srcs = [":hello_proto"],
+    use_well_known_protos = True
+)
+objc_grpc_lib(
+    name = "hello_grpc_objc",
+    srcs = [":hello_proto"],
+    deps = [":hello_proto_objc"],
+    use_well_knonw_protos = True
+)
+# app-specific library below
+objc_library(
+    name = "HelloWorldLibrary",
+    ...
+    deps = [
+        ":hello_proto_objc",
+        ":hello_grpc_objc",
+        "@com_github_grpc_grpc//:gRPCCertificates"
+    ]
+)
+ios_application(
+    ...
+    deps = [":HelloWorldLibrary"]
+)
+```
+Import the `hello.proto` stubs like:
+```
+#import <HelloWorld/Hello.pbobjc.h>
+```
+
+
```

Comment:
Unrelated to this doc, but I think one example (or build test) needs to test when one proto file imports another proto file in another package or workspace

Reply:
@muxi 
Well, this is to kinda illustrate the one of the open issues below... But I will write up a more comprehensive example!

Reply:
BTW I've renamed the file to L56-XXX